rust/connlib/clients/shared/src/control.rs

```diff
@@ -100,15 +100,16 @@ impl<CB: Callbacks + 'static> ControlPlane<CB> {
                     return Err(e);
                 } else {
                     *init = true;
-                    *self.fallback_resolver.lock() =
-                        create_resolver(interface.upstream_dns, self.tunnel.callbacks());
                     tracing::info!("Firezone Started!");
                 }
             } else {
                 tracing::info!("Firezone reinitializated");
             }
         }
 
+        self.tunnel.set_upstream_dns(&interface.upstream_dns);
+        *self.fallback_resolver.lock() =
```

Comment:
Blocking locks are used in an async fn here because the critical section is short and there are no nested locks, so it's unlikely to block for long or deadlock, right?

Reply:
The important thing is that we don't need to hold the lock across an `await` point, given that, it's just less performant to use an async one.
[ref](https://tokio.rs/tokio/tutorial/shared-state#on-using-stdsyncmutex)

Reply:
Okay. The Tokio docs were a little confusing to me, saying both "it's not better because the async mutex is sync internally" but also "async locking doesn't block"

But I think this clears it up - "The primary use case for the async mutex is to provide shared mutable access to IO resources such as a database connection. If the value behind the mutex is just data, it’s usually appropriate to use a blocking mutex such as the one in the standard library or [parking_lot](https://docs.rs/parking_lot)."

Cause if it's just data then the worst that can happen is that the thread could be paused while holding the lock, but if the thread is paused, it wouldn't be executing any other tasks anyway. And in both cases, any other task needing the lock will block.